# Numpy 예제

In [1]:
import numpy as np

**예제 1. np.arange()과 reshape() 메소드를 이용해 아래와 같은 array(matrix)를 생성하시오**

a = array([[ 1,  4,  7, 10],<br>
          [13, 16, 19, 22],<br>
          [25, 28, 31, 34]])

In [2]:
a = np.arange(1, 35, 3).reshape(3, 4)
a

array([[ 1,  4,  7, 10],
       [13, 16, 19, 22],
       [25, 28, 31, 34]])

**예제 2. 위에서 생성한 array의 행의 개수(num_row)와 열의 개수(num_col)를 추출하시오**

In [3]:
num_row = a.shape[0]
num_col = a.shape[1]
print(num_row, num_col)

3 4


In [4]:
num_row = len(a)
num_col = len(a.T)
print(num_row, num_col)

3 4


**예제 3. 두 matrix에 대해서 요소별로 각각 큰 값을 저장하는 array(matrix)를 생성하시오**

예)

두 array에 대하여

array([[1 , 2],<br>
       [4, 5]])<br>

array([[2 , 1],<br>
       [3, 7]])<br>
       
**-> result:**<br>
array([[2, 2],<br>
       [4, 7]])<br>

In [5]:
## test case ##
np.random.seed(0)
a = np.random.random((3, 3))
b = np.random.random((3, 3))
print('a:', a)
print('b:', b)

a: [[0.5488135  0.71518937 0.60276338]
 [0.54488318 0.4236548  0.64589411]
 [0.43758721 0.891773   0.96366276]]
b: [[0.38344152 0.79172504 0.52889492]
 [0.56804456 0.92559664 0.07103606]
 [0.0871293  0.0202184  0.83261985]]


In [6]:
# return 방식
np.maximum(a, b)

array([[0.5488135 , 0.79172504, 0.60276338],
       [0.56804456, 0.92559664, 0.64589411],
       [0.43758721, 0.891773  , 0.96366276]])

In [7]:
# In-place 방식
masking = a < b
a[masking] = b[masking]
a

array([[0.5488135 , 0.79172504, 0.60276338],
       [0.56804456, 0.92559664, 0.64589411],
       [0.43758721, 0.891773  , 0.96366276]])

array([[0.5488135 , 0.79172504, 0.60276338],<br>
       [0.56804456, 0.92559664, 0.64589411],<br>
       [0.43758721, 0.891773  , 0.96366276]])

**예제 4. 예제3에서 구현한 코드를 이용하여 어떤 matrix에 대해 각각의 값(요소)이 음수일 때는 0, 양수일 때는 그 값을 그대로 저장하는 코드를 구현하시오**

예)

a = array([[ 1.76405235,  0.40015721,  0.97873798],<br>
       [ 2.2408932 ,  1.86755799, -0.97727788],<br>
       [ 0.95008842, -0.15135721, -0.10321885]])
       
**-> result:**<br>
array([[1.76405235, 0.40015721, 0.97873798],<br>
       [2.2408932 , 1.86755799, 0.        ],<br>
       [0.95008842, 0.        , 0.        ]])

In [8]:
np.random.seed(0)
a = np.random.randn(5, 5)  # np.random.randn은 argument로 튜플을 받지 않음에 주의
print(a)

[[ 1.76405235  0.40015721  0.97873798  2.2408932   1.86755799]
 [-0.97727788  0.95008842 -0.15135721 -0.10321885  0.4105985 ]
 [ 0.14404357  1.45427351  0.76103773  0.12167502  0.44386323]
 [ 0.33367433  1.49407907 -0.20515826  0.3130677  -0.85409574]
 [-2.55298982  0.6536186   0.8644362  -0.74216502  2.26975462]]


In [9]:
# return 방식
np.maximum(0, a)

array([[1.76405235, 0.40015721, 0.97873798, 2.2408932 , 1.86755799],
       [0.        , 0.95008842, 0.        , 0.        , 0.4105985 ],
       [0.14404357, 1.45427351, 0.76103773, 0.12167502, 0.44386323],
       [0.33367433, 1.49407907, 0.        , 0.3130677 , 0.        ],
       [0.        , 0.6536186 , 0.8644362 , 0.        , 2.26975462]])

In [10]:
# In-place 방식
masking = a < 0
a[masking] = 0
a

array([[1.76405235, 0.40015721, 0.97873798, 2.2408932 , 1.86755799],
       [0.        , 0.95008842, 0.        , 0.        , 0.4105985 ],
       [0.14404357, 1.45427351, 0.76103773, 0.12167502, 0.44386323],
       [0.33367433, 1.49407907, 0.        , 0.3130677 , 0.        ],
       [0.        , 0.6536186 , 0.8644362 , 0.        , 2.26975462]])

array([[1.76405235, 0.40015721, 0.97873798, 2.2408932 , 1.86755799],<br>
       [0.        , 0.95008842, 0.        , 0.        , 0.4105985 ],<br>
       [0.14404357, 1.45427351, 0.76103773, 0.12167502, 0.44386323],<br>
       [0.33367433, 1.49407907, 0.        , 0.3130677 , 0.        ],<br>
       [0.        , 0.6536186 , 0.8644362 , 0.        , 2.26975462]])

**예제 5. 어떤 matrix에 대해 각 요소가 0.5 이상이면 1, 0.5 이하면 0을 저장하는 코드를 구현하시오**

예)

a = array([[0.94374808, 0.6818203 ],<br>
       [0.3595079 , 0.43703195]])<br>
       
**-> result:**<br>
array([[1., 1.],<br>
       [0., 0.]])

In [11]:
np.random.seed(0)
a = np.random.random((3, 4))
print(a)

[[0.5488135  0.71518937 0.60276338 0.54488318]
 [0.4236548  0.64589411 0.43758721 0.891773  ]
 [0.96366276 0.38344152 0.79172504 0.52889492]]


In [12]:
# return 방식
np.round(a)

array([[1., 1., 1., 1.],
       [0., 1., 0., 1.],
       [1., 0., 1., 1.]])

In [13]:
# In-place 방식
a[a>0.5] = 1
a[a<=0.5] = 0
a

array([[1., 1., 1., 1.],
       [0., 1., 0., 1.],
       [1., 0., 1., 1.]])

array([[1., 1., 1., 1.],<br>
       [0., 1., 0., 1.],<br>
       [1., 0., 1., 1.]])

**예제 6. 아래와 같은 matrix생성 후 np.sum을 이용하여 각 열들의 합과 각 행들의 합을 구하시오**

a = np.arange(1, 10).reshape(3, 3)

각 열들의 합:<br>
array([[ 6],<br>
       [15],<br>
       [24]])<br>
       
각 행들의 합:<br>
array([[12, 15, 18]])

In [14]:
a = np.arange(1, 10).reshape(3, 3)
a

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [15]:
np.sum(a, axis=1, keepdims=True)

array([[ 6],
       [15],
       [24]])

In [16]:
np.sum(a, axis=0, keepdims=True)

array([[12, 15, 18]])

**예제 7. np.concatenate()를 이용하여 아래와 같은 결과를 보이시오**

a = array([[1, 2], [3, 4]])<br>
b = array([[5, 6], [7, 8]])<br>
c = array([[9, 10]])

**Result:**

array([[1, 2],<br>
       [3, 4],<br>
       [5, 6],<br>
       [7, 8]])

array([[1, 2, 5, 6],<br>
       [3, 4, 7, 8]])

array([[ 1,  2,  9],<br>
       [ 3,  4, 10]])

In [18]:
a = np.array([[1, 2], [3, 4]])
b = np.array([[5, 6], [7, 8]])
c = np.array([[9, 10]])

In [19]:
np.concatenate((a, b), axis=0)

array([[1, 2],
       [3, 4],
       [5, 6],
       [7, 8]])

In [20]:
np.concatenate((a, b), axis=1)

array([[1, 2, 5, 6],
       [3, 4, 7, 8]])

In [21]:
np.concatenate((a, c.T), axis=1)

array([[ 1,  2,  9],
       [ 3,  4, 10]])